src/cell.cpp

```diff
@@ -635,6 +635,93 @@ void DAGCell::to_hdf5(hid_t group_id) const { return; }
 
 #endif
 
+//==============================================================================
+// UniversePartitioner implementation
+//==============================================================================
+
+UniversePartitioner::UniversePartitioner(const Universe& univ)
+{
+  // Find all of the z-planes in this universe.  Add them to the surfs_ member.
+  // Use surf_set for O(1) searches that ensure surfs_ entries are not repeated.
+  std::unordered_set<int32_t> surf_set;
+  for (auto i_cell : univ.cells_) {
+    for (auto token : model::cells[i_cell]->rpn_) {
+      if (token < OP_UNION) {
+        auto i_surf = std::abs(token) - 1;
+        const auto* surf = model::surfaces[i_surf].get();
+        if (const auto* zplane = dynamic_cast<const SurfaceZPlane*>(surf)) {
+          if (surf_set.find(i_surf) == surf_set.end()) {
+            surf_set.insert(i_surf);
+            surfs_.push_back(i_surf);
+          }
+        }
+      }
+    }
+  }
+
+  // Define a functor for comparing z-planes by their location, and use it to
+  // sort the surfs_ member.
+  struct {
+    bool operator()(int32_t i_surf, int32_t j_surf) const
+    {
+      const auto* surf = model::surfaces[i_surf].get();
+      const auto* zplane = dynamic_cast<const SurfaceZPlane*>(surf);
+      double zi = zplane->z0_;
+      surf = model::surfaces[j_surf].get();
+      zplane = dynamic_cast<const SurfaceZPlane*>(surf);
+      double zj = zplane->z0_;
+      return zi < zj;
+    }
+  } compare_surfs;
+  std::sort(surfs_.begin(), surfs_.end(), compare_surfs);
+
+  // Populate the partition lists.
+  partitions_.resize(surfs_.size() + 1);
+  for (auto i_cell : univ.cells_) {
+    // Find the tokens for bounding z-planes.
+    int32_t min_token = 0, max_token = 0;
+    double min_z, max_z;
```

Comment:
I think we should make an exception to ES.10 for cases like this.  The ES.10 example has a bunch of arrays which is certainly confusing, but here the types are super simple.  I think adding a on of separate lines to declare blocks of simple variables like these hurts rather than helps readability.

Reply:
Yeah, I don't feel too strongly about it for this case, so I'm fine leaving as is